payment-guarantee/src/contract.rs

```diff
@@ -0,0 +1,698 @@
+use std::borrow::BorrowMut;
+
+use cosmwasm_std::{attr, Binary, BlockInfo, Coin, coin, Deps, DepsMut, Empty, Env, from_binary, MessageInfo, Order, Pair, Response, StdError, StdResult, to_binary};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cw721::{Cw721QueryMsg, Cw721ReceiveMsg, Expiration, NftInfoResponse, NumTokensResponse, TokensResponse};
+use cw721_base::ContractError;
+use cw721_base::msg::{InstantiateMsg, MinterResponse};
+
+use crate::msg::{ExecuteMsg, MintMsg, QueryMsg};
+use crate::query::{NftInfoWithPriceResponse, TokenPriceResponse};
+use crate::receiver::NftInfoWithPriceMsg;
+use crate::state::TOKEN_PRICE;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> StdResult<Response> {
+    cw721_base::contract::instantiate(deps, _env, _info, msg)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Mint(msg) => execute_mint(deps, env, info, msg),
+        ExecuteMsg::Approve {
+            spender,
+            token_id,
+            expires,
+        } => cw721_base::contract::execute_approve(deps, env, info, spender, token_id, expires),
+        ExecuteMsg::Revoke {
+            spender,
+            token_id,
+        } => cw721_base::contract::execute_revoke(deps, env, info, spender, token_id),
+        ExecuteMsg::ApproveAll {
+            operator,
+            expires,
+        } => cw721_base::contract::execute_approve_all(deps, env, info, operator, expires),
+        ExecuteMsg::RevokeAll {
+            operator
+        } => cw721_base::contract::execute_revoke_all(deps, env, info, operator),
+        ExecuteMsg::TransferNft {
+            recipient,
+            token_id,
+        } => cw721_base::contract::execute_transfer_nft(deps, env, info, recipient, token_id),
+        ExecuteMsg::SendNft {
+            contract,
+            token_id,
+        } => execute_send_nft(deps, env, info, contract, token_id),
+        ExecuteMsg::ReceiveNft(msg) => execute_receive_nft(deps, env, info, msg),
+    }
+}
+
+pub fn execute_mint(
+    mut deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: MintMsg,
+) -> Result<Response, ContractError> {
+    let nft_mint_msg = cw721_base::msg::MintMsg {
+        token_id: msg.token_id.clone(),
+        owner: msg.owner.clone(),
+        name: msg.name.clone(),
+        description: msg.description.clone(),
+        image: msg.image.clone(),
+    };
+
+    let mint_response = cw721_base::contract::execute_mint(deps.branch(), _env, info.clone(), nft_mint_msg.clone());
+
+    if mint_response.is_err() {
+        return mint_response;
+    }
+
+    TOKEN_PRICE.save(deps.storage, msg.token_id.clone(), &msg.price);
```

Comment:
The price of the token is stored in a separate repository based on the token_id.
This can also be changed to be stored in the description of the token.